kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java

```diff
@@ -81,6 +83,113 @@ void roundTripSingleRecord(KafkaCluster cluster, Topic topic, TestKmsFacade<?, ?
+                var consumer = tester.consumer(Map.of(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest",
+                        ConsumerConfig.GROUP_ID_CONFIG, UUID.randomUUID().toString(),
+                        ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed"))) {
+            producer.initTransactions();
+            // send to the same partition to demonstrate a message appended to the same partition after the abort is made available
+            String key = "key";
+            withTransaction(producer, transactionProducer -> {
+                producer.send(new ProducerRecord<>(topic.name(), key, "aborted message")).get(5, TimeUnit.SECONDS);
+            }).abortTransaction();
+
+            withTransaction(producer, transactionProducer -> {
+                producer.send(new ProducerRecord<>(topic.name(), key, HELLO_WORLD)).get(5, TimeUnit.SECONDS);
+            }).commitTransaction();
+
+            consumer.subscribe(List.of(topic.name()));
+            var records = consumer.poll(Duration.ofSeconds(2));
+            assertThat(records.iterator())
+                    .toIterable()
+                    .singleElement()
+                    .extracting(ConsumerRecord::value)
+                    .isEqualTo(HELLO_WORLD);
+        }
+    }
+
+    // check that records from uncommitted transaction are not exposed to read_committed clients
+    @TestTemplate
+    void roundTripTransactionalIsolation(KafkaCluster cluster, Topic topic, TestKmsFacade<?, ?, ?> testKmsFacade) {
+        var testKekManager = testKmsFacade.getTestKekManager();
+        testKekManager.generateKek(topic.name());
+
+        var builder = proxy(cluster);
+
+        builder.addToFilters(buildEncryptionFilterDefinition(testKmsFacade));
+
+        try (var tester = kroxyliciousTester(builder);
+                var producer = tester.producer(Map.of(ProducerConfig.TRANSACTIONAL_ID_CONFIG, UUID.randomUUID().toString()));
+                var consumer = tester.consumer(Map.of(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest",
+                        ConsumerConfig.GROUP_ID_CONFIG, UUID.randomUUID().toString(),
+                        ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed"))) {
+            producer.initTransactions();
+
+            withTransaction(producer, transactionProducer -> {
+                transactionProducer.send(new ProducerRecord<>(topic.name(), "uncommitted message")).get(5, TimeUnit.SECONDS);
+            });
+
+            consumer.subscribe(List.of(topic.name()));
+            var records = consumer.poll(Duration.ofSeconds(2));
+            assertThat(records.iterator())
+                    .isExhausted();
+        }
+    }
+
+    interface ExceptionalConsumer<T> {
+        void accept(T t) throws Exception;
+    }
+
+    <K, V> Producer<K, V> withTransaction(Producer<K, V> producer, ExceptionalConsumer<Producer<K, V>> consumer) {
+        producer.beginTransaction();
+        try {
+            consumer.accept(producer);
```

Comment:
seeing `consumer` in this context shocks my brain because of the collision with KafkaConsumer. Rename var to `action`.